stan/math/rev/mat/fun/cov_exp_quad.hpp

```diff
@@ -0,0 +1,305 @@
+              for (size_t i = 0; i < static_cast<size_t>(size_); ++i)
+                cov_diag_[i] = new vari(sigma_sq_d_, false);
+          }
+
+        virtual void chain() {
+          double adjl(0.0);
+
+          for (size_t i = 0; i < static_cast<size_t>(size_vech_); ++i) {
+            vari* el_low = cov_lower_[i];
+            vari* el_high = cov_upper_[i];
+            double prod_adj_cov_low = el_low->adj_ * el_low->val_;
+            double prod_adj_cov_high = el_high->adj_ * el_high->val_;
+            double prod_add = prod_adj_cov_low + prod_adj_cov_high;
+            adjl += prod_add * dist_[i];
+          }
+          l_vari_->adj_ +=  adjl / std::pow(l_d_, 3);
+        }
+    };
+
+    /**
+     * Returns a squared exponential kernel.
+     *
+     * @param x std::vector input that can be used in square distance
+     *    Assumes each element of x is the same size
+     * @param sigma standard deviation
+     * @param l length scale
+     * @return squared distance
+     * @throw std::domain_error if sigma <= 0, l <= 0, or
+     *   x is nan or infinite
+     */
+    inline
+    Eigen::Matrix<var, -1, -1>
+    cov_exp_quad(const std::vector<double>& x,
+                 const var& sigma,
+                 const var& l) {
+      check_positive("cov_exp_quad", "sigma", sigma);
+      check_positive("cov_exp_quad", "l", l);
+      for (size_t n = 0; n < x.size(); n++)
+        check_not_nan("cov_exp_quad", "x", x[n]);
+
+      int x_size = x.size();
+      Eigen::Matrix<var, Eigen::Dynamic, Eigen::Dynamic>
+        cov(x_size, x_size);
+      if (x_size == 0)
+        return cov;
+
+      cov_exp_quad_vari<double, var, var> *baseVari
+        = new cov_exp_quad_vari<double, var, var>(x, sigma, l);
+
+      size_t pos = 0;
+      for (size_t j = 0; j < static_cast<size_t>(x_size - 1); ++j) {
+        for (size_t i = (j + 1); i < static_cast<size_t>(x_size); ++i) {
+          cov.coeffRef(i, j).vi_ = baseVari->cov_lower_[pos];
+          cov.coeffRef(j, i).vi_ = baseVari->cov_upper_[pos];
+          ++pos;
+        }
+        cov.coeffRef(j, j).vi_ = baseVari->cov_diag_[j];
+      }
+      cov.coeffRef(x_size - 1, x_size - 1).vi_
+        = baseVari->cov_diag_[x_size - 1];
+      return cov;
+    }
+    /**
+     * Returns a squared exponential kernel.
+     *
+     * @param x std::vector input that can be used in square distance
+     *    Assumes each element of x is the same size
+     * @param sigma standard deviation
+     * @param l length scale
+     * @return squared distance
+     * @throw std::domain_error if sigma <= 0, l <= 0, or
+     *   x is nan or infinite
+     */
+    inline
+    Eigen::Matrix<var, -1, -1>
+    cov_exp_quad(const std::vector<double>& x,
+                 double sigma,
+                 const var& l) {
+      check_positive("cov_exp_quad", "sigma", sigma);
+      check_positive("cov_exp_quad", "l", l);
+      for (size_t n = 0; n < x.size(); n++)
+        check_not_nan("cov_exp_quad", "x", x[n]);
```

Comment:
OK, missed that.  So skip that comment!  In which case, I'll go with my first comment, which is that there should be vector vesions of these tests that will let the user know which index is off rather than just saying "x".